### PR TITLE
perf: improve computeRange performance

### PR DESCRIPTION
**Pull request by FezVrasta**
### PR and Code Checklist

- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Context

The existing `createRangeHelper` and `computeRange` functions trigger the garbage collector a lot and slow down the whole operation.

### Changes

A new `fastComputeRange` function has been added, that provides a 3x speed improvement over the existing function. While not triggering a single GC action.

~I would like to completely remove the slow `createRangeHelper` but I need to better understand how it's used first. Some help would be apreciated.~

- [x] Documentation and TypeScript definitions were updated to match those changes

### Results

**Before**
CleanShot 2022-02-09 at 5 49 14@2x
CleanShot 2022-02-09 at 6 01 45@2x

**After**
CleanShot 2022-02-09 at 5 43 55@2x

### Testing

- [ ] This change adds or fixes unit tests
- [ ] All tests complete without errors on the following environment:
  - **vtk.js**: 
  - **OS**: 
  - **Browser**: 

_This contribution is funded by [Sirona Medical](https://sironamedical.com)._

### CODE REVIEW

**Comment by jourdain** [comment]
LGTM

File: Sources/Common/Core/DataArray/index.js [hunk 1]
```diff
@@ -9,6 +9,31 @@ const TUPLE_HOLDER = [];
...
+  let max;
+  let x;
+  let i;
+
+  if (len === 0) {
+    return null;
+  }
+  min = arr[0];
+  max = min;
+  for (i = 1; i < len; i++) {
```
Comment (jadh4v) [inline Sources/Common/Core/DataArray/index.js:26]: I think you will have to pass `offset` and `numberOfComponents` as arguments and do `i=offset` and `i += numberOfComponents` to properly handle multi-channel data.

Comment (FezVrasta) [reply Sources/Common/Core/DataArray/index.js:26]: I sent a commit that _should_ address multi-channel data too, may you take a look please?

**Review by sankhesh — APPROVED** [review]
LGTM

File: Sources/Common/Core/DataArray/index.js [hunk 2]
```diff
@@ -36,29 +64,24 @@ function createRangeHelper() {
...
 function computeRange(values, component = 0, numberOfComponents = 1) {
-  const helper = createRangeHelper();
-  const size = values.length;
-  let value = 0;
-
   if (component < 0 && numberOfComponents > 1) {
     // Compute magnitude
+    const size = values.length;
+    let value = 0;
+    const data = [];
```
Comment (finetjul) [inline Sources/Common/Core/DataArray/index.js:71]: you could allocate the final size of the array instead of "push"

Comment (FezVrasta) [reply Sources/Common/Core/DataArray/index.js:71]: I don't quite get how this part is working. Care to send a patch?

Comment (finetjul) [reply Sources/Common/Core/DataArray/index.js:71]:
```
const data = [];
data.length = size / numberOfComponents;
...
  data[i/numberOfComponents] = value;
```
(you may have to have `i` ranging from 0 to size / numberofComponents to reduce the number of operations).
e.g.
```
const numberOfValues = size / numberOfComponents)
const data = new Float64Array(numberOfValues);
for (let i = 0, j = 0; i < numberOfValues; ++i) {
  for (let nextJ = j + numberOfComponents; j < nextJ; ++j) {
     data[i] += values[j]*values[j];
   }
   data[i] *= 0.5;
}
```

Comment (FezVrasta) [reply Sources/Common/Core/DataArray/index.js:71]: Couldn't this lead to non-integer numbers as `numberOfValues`?

Comment (finetjul) [reply Sources/Common/Core/DataArray/index.js:71]: it should never be the case.

Comment (FezVrasta) [reply Sources/Common/Core/DataArray/index.js]: I pushed the change.

Comment (finetjul) [reply Sources/Common/Core/DataArray/index.js]: Is it any faster ? 
if not, then it may not worth it.

Comment (FezVrasta) [reply Sources/Common/Core/DataArray/index.js]: Your version is 25% faster

File: Sources/Common/Core/DataArray/index.js [hunk 3]
```diff
@@ -9,6 +9,34 @@ const TUPLE_HOLDER = [];
...
+  let max;
+  let x;
+  let i;
+
+  if (len === 0) {
+    return null;
+  }
+  min = arr[0];
+  max = min;
+  for (i = 1; i < len; i++) {
```
Comment (jadh4v) [inline Sources/Common/Core/DataArray/index.js:26]:
```suggestion
  for (i = offset; i < len; i += numOfComponents) {
```

File: Sources/Common/Core/DataArray/index.js [hunk 4]
```diff
@@ -9,6 +9,34 @@ const TUPLE_HOLDER = [];
 // Global methods
 // ----------------------------------------------------------------------------
 
+// Original source from https://www.npmjs.com/package/compute-range
+// Modified to accept type arrays
+function fastComputeRange(arr) {
```
Comment (jadh4v) [inline Sources/Common/Core/DataArray/index.js:14]:
```suggestion
function fastComputeRange(arr, offset, numOfComponents) {
```

File: Sources/Common/Core/DataArray/index.js [hunk 5]
```diff
@@ -36,29 +64,21 @@ function createRangeHelper() {
...
-      helper.add(value);
+      data[i] **= 0.5;
     }
-    return helper.getRange();
-  }
-
-  const offset = component < 0 ? 0 : component;
-  for (let i = offset; i < size; i += numberOfComponents) {
-    helper.add(values[i]);
+    return fastComputeRange(data);
```
Comment (jadh4v) [inline Sources/Common/Core/DataArray/index.js:78]:
```suggestion
    return fastComputeRange(data, 0, 1);
```

File: Sources/Common/Core/DataArray/index.js [hunk 6]
```diff
@@ -36,29 +64,21 @@ function createRangeHelper() {
...
   }
 
-  return helper.getRange();
+  return fastComputeRange(values);
```
Comment (jadh4v) [inline Sources/Common/Core/DataArray/index.js:81]:
```suggestion
  return fastComputeRange(values, component < 0 ? 0 : component, numberOfComponents);
```

**Review by sankhesh — APPROVED** [review]

**Comment by FezVrasta** [comment]
Any ETA to get this released?

**Comment by finetjul** [comment]
@FezVrasta would you mind writing a unit test (and squashing your commits to keep only 2: one from yours and one from @jadh4v ?

@jourdain how about removing `createRangeHelper()`, or can it be flagged as obsolete ?

**Comment by FezVrasta** [comment]
@finetjul don't you already have some unit tests for publicAPI.getRange? I'd defer the unit test creation to someone else since I don't completely understand how the most advanced modes of this utility work.

**Comment by jourdain** [comment]
If `createRangeHelper()` is not needed anymore, we can surely remove it, unless it was public and used outside of this context.

**Comment by FezVrasta** [comment]
`createRangeHelper` is exposed to `STATIC` so I suppose someone may be using it in userland?

**Comment by FezVrasta** [comment]
@jourdain @jadh4v I experimented with an alternative approach and I noticed that `Array.prototype.sort` is approx 20% faster than the algorithm this PR introduced, probably the fact the built-in sorth algorithm is ran as native code gives it an additional perf boost.

This can only work when `numberOfComponents` is set to `1` unfortunately, but I think it may be worth it implementing this change?

Here's the benchmarks:
https://jsbench.me/kokzo24f6t/1

**Comment by jourdain** [comment]
What about an array with NaN

**Pull request merged**